src/win32/path_w32.c

```diff
@@ -303,3 +308,230 @@ char *git_win32_path_8dot3_name(const char *path)
+	dir = git__calloc(1, sizeof(DIR));
+	dir->h = FindFirstFileExW(pathw, FindExInfoBasic, &dir->f, FindExSearchNameMatch, NULL, 0);
+	dir->first = 1;
+	if (dir->h == INVALID_HANDLE_VALUE)
+	{
+		error = -2;
+		giterr_set(GITERR_OS, "Could not open directory '%s'", path);
+		goto clean_up_and_exit;
+	}
+	
+	if (repo_path_len > PATH__MAX_UNC_LEN)
+	{
+		error = -3;
+		giterr_set(GITERR_OS, "Could not open directory '%s'", path);
+		goto clean_up_and_exit;
+	}
+
+	memcpy(work_path, repo_path, repo_path_len);
+
+	while (dir)
+	{
+		if (!git_path_is_dot_or_dotdotW(dir->f.cFileName)) {
+			path_len = git__utf16_to_8(work_path + repo_path_len, ARRAYSIZE(work_path) - repo_path_len, dir->f.cFileName);
+
+			work_path[path_len + repo_path_len] = '\0';
+			path_len = path_len + repo_path_len;
+
+			cmp_len = min(start_len, path_len);
+			if (!(cmp_len && strncomp(work_path, start_stat, cmp_len) < 0)) {
+				cmp_len = min(end_len, path_len);
+				if (!(cmp_len && strncomp(work_path, end_stat, cmp_len) > 0)) {
+					fMode = S_IREAD;
+
+					if (dir->f.dwFileAttributes & FILE_ATTRIBUTE_DIRECTORY)
+						fMode |= S_IFDIR;
+					else
+						fMode |= S_IFREG;
+
+					if (!(dir->f.dwFileAttributes & FILE_ATTRIBUTE_READONLY))
+						fMode |= S_IWRITE;
+
+					ps = git__calloc(1, sizeof(git_path_with_stat) + path_len + 2);
+					memcpy(ps->path, work_path, path_len + 1);
+					ps->path_len = path_len;
+					ps->st.st_atime = filetime_to_time_t(&dir->f.ftLastAccessTime);
+					ps->st.st_ctime = filetime_to_time_t(&dir->f.ftCreationTime);
+					ps->st.st_mtime = filetime_to_time_t(&dir->f.ftLastWriteTime);
+					ps->st.st_size = dir->f.nFileSizeHigh;
+					ps->st.st_size <<= 32;
+					ps->st.st_size |= dir->f.nFileSizeLow;
+					ps->st.st_dev = ps->st.st_rdev = (_getdrive() - 1);
+					ps->st.st_mode = (mode_t)fMode;
+					ps->st.st_ino = 0;
+					ps->st.st_gid = 0;
+					ps->st.st_uid = 0;
+					ps->st.st_nlink = 1;
+
+					if (dir->f.dwFileAttributes & FILE_ATTRIBUTE_REPARSE_POINT) {
+						if (git_win32_path_readlink_w(target, dir->f.cFileName) >= 0) {
+							ps->st.st_mode = (ps->st.st_mode & ~S_IFMT) | S_IFLNK;
+
+							/* st_size gets the UTF-8 length of the target name, in bytes,
+							*  not counting the NULL terminator */
+							if ((ps->st.st_size = git__utf16_to_8(NULL, 0, target)) < 0) {
+								error = -4;
+								giterr_set(GITERR_OS, "Could not manage reparse link '%s'", dir->f.cFileName);
+								goto clean_up_and_exit;
+							}
+						}
+					}
+
+					if (S_ISDIR(ps->st.st_mode)) {
+						ps->path[ps->path_len++] = '/';
+						ps->path[ps->path_len] = '\0';
+					} else if (!S_ISREG(ps->st.st_mode) && !S_ISLNK(ps->st.st_mode)) {
+						git__free(ps);
+						ps = NULL;
+					}
+
+					if (ps)
+					{
+						git_vector_insert(contents, ps);
+					}
+				}
+			}
+		}
+
+		memset(&dir->f, 0, sizeof(git_path_with_stat));
+		dir->first = 0;
+
+		if (!FindNextFileW(dir->h, &dir->f))
+		{
+			if (GetLastError() == ERROR_NO_MORE_FILES)
```

Comment:
Please add braces around the `break` here, eg:

    if (GetLastError() == ERROR_NO_MORE_FILES) {
        break;
    } else {
        error = -1;
        ...